03-Binary-Search-Trees/02-Binary-Search-Trees.md

```diff
@@ -191,480 +86,938 @@ class Tree:
+        self.assertEqual("Char", t.root.right.left.value)
+        self.assertEqual(15, t.root.right.right.key)
+        self.assertEqual("Ada", t.root.right.right.value)
+        self.assertEqual(25, t.root.right.right.right.key)
+        self.assertEqual("Kari", t.root.right.right.right.value)
+```
 
-##### !end-answer
+##### !end-tests
 
 <!-- other optional sections -->
-##### !hint
+##### !hint 
+The `add` method is similar to the the `find` method - but your new node will always be the child of a node that is currently a leaf in the existing tree!
 
-What's the worst case?
+Try identifying your base and recursive cases in the iterative solution you just wrote.
 
-##### !end-hint
-<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
-##### !explanation
+Still feeling stuck? Check this video walkthrough of the solution.
 
-Worst-case you have to travel to the last node in the chain and since there are 5 levels, it takes 5 comparisons
+<!-- ADD VIDEO WALKTHROUGH -->
+##### !end-hint
+##### !explanation 
+An example of a working implementation:
+```python
+    # helper function handles the recursion
+    def add_helper(self, current_node, new_node):
+
+        # if new node should be in left subtree (less than current node)
+        if new_node.key < current_node.key: 
+            # if current node is a leaf
+            if not current_node.left:
+                # make new node left child of current node
+                current_node.left = new_node
+                return
+            # Otherwise, recurse through left subtree of current node
+            self.add_helper(current_node.left, new_node)
+        # if new node should be in right subtree 
+        # (greater than or equal to current node)
+        else:
+            # if current node is a leaf
+            if not current_node.right:
+                # make new node right child of current node
+                current_node.right = new_node
+                return
+            # Otherwise, recurse through right subtree of current node
+            self.add_helper(current_node.right, new_node)
+
+    def add(self, key, value = None):
+        # If tree is empty
+        if not self.root:
+            # Make new node the root
+            self.root = TreeNode(key, value)
+        # Otherwise, initiate tree traversal by calling add on root
+        else:
+            new_node = TreeNode(key, value)
+            self.add_helper(self.root, new_node)
+```
 
 ##### !end-explanation
 
 ### !end-challenge
-
 <!-- ======================= END CHALLENGE ======================= -->
 
+
+### Deletion
+Deleting a node froma binary search tree is a little more complex than either searching for or inserting a node, because we need to ensure that when we remove the node the resulting tree still maintains the property that all left children are smaller than their parent nodes and all right children are larger than their parent nodes.
```

Comment:
```suggestion
Deleting a node from a binary search tree is a little more complex than either searching for or inserting a node, because we need to ensure that when we remove the node the resulting tree still maintains the property that all left children are smaller than their parent nodes and all right children are larger than their parent nodes.
```